hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/CassAlertsServiceImpl.java

```diff
@@ -0,0 +1,849 @@
+                session.execute(insertAlertTrigger.bind(a.getAlertId(), a.getTriggerId()));
+                session.execute(insertAlertCtime.bind(a.getAlertId(), a.getCtime()));
+                session.execute(insertAlertStatus.bind(a.getAlertId(), a.getStatus().name()));
+            }
+        } catch (Exception e) {
+            msgLog.errorDatabaseException(e.getMessage());
+            throw e;
+        }
+    }
+
+    public List<Alert> getAlerts(AlertsCriteria criteria) throws Exception {
+        if (session == null) {
+            throw new RuntimeException("Cassandra session is null");
+        }
+        boolean filter = (null != criteria && criteria.hasCriteria());
+
+        if (filter) {
+            log.debugf("getAlerts criteria: %s", criteria.toString());
+        }
+
+        List<Alert> alerts = new ArrayList<>();
+        Set<String> alertIds = new HashSet<>();
+
+        StringBuilder sAlerts = new StringBuilder("SELECT payload FROM ").append(keyspace).append(".alerts ");
+
+        if (filter) {
+            /*
+                Triggers id can be passed explicitly in the criteria or indirectly via tags.
+             */
+            Set<String> triggerIds = new HashSet<>();
+            if (isEmpty(criteria.getTriggerIds())) {
+                if (!isEmpty(criteria.getTriggerId())) {
+                    triggerIds.add(criteria.getTriggerId());
+                }
+            } else {
+                for (String triggerId : criteria.getTriggerIds()) {
+                    if (isEmpty(triggerId)) {
+                        continue;
+                    }
+                    triggerIds.add(triggerId);
+                }
+            }
+            if (!isEmpty(criteria.getTags()) || criteria.getTag() != null) {
+                Set<Tag> tags = new HashSet<>();
+                if (criteria.getTags() != null)  {
+                    tags.addAll(criteria.getTags());
+                }
+                Tag tag = criteria.getTag();
+                if (tag != null) {
+                    tags.add(tag);
+                }
+                triggerIds.addAll(getTriggersIdByTags(tags));
+            }
+
+            /*
+                Get alertIds filtered by triggerIds
+             */
+            Set<String> alertIdsFilteredByTriggers = new HashSet<>();
+            boolean filterByTriggers = false;
+            if (triggerIds.size() > 0) {
+                filterByTriggers = true;
+                StringBuilder sAlertsTriggers = new StringBuilder("SELECT alertId FROM ").append(keyspace)
+                        .append(".alerts_triggers WHERE ");
+                if (triggerIds.size() == 1) {
+                    sAlertsTriggers.append("( triggerId = '");
+                    sAlertsTriggers.append(triggerIds.iterator().next());
+                    sAlertsTriggers.append("' )");
+                } else {
+                    sAlertsTriggers.append("( triggerId IN (");
```

Comment:
This is good to know.  I suppose "criteria-style" querying is not a great an idea against cassandra as it was against an RDB.  One problem we'd have trying to convert to multiple async queries would be the potential for a combinatorial issue when dealing with multiple "IN clause" criteria.  I'm not sure what the solution is here.  We could leave as-is and see how it goes, or we could be proactive and try to make a change, possibly changing the API to allow for more restricted querying.  But I can certainly imagine the need for clients to want to query for alerts for say, for multiple triggers (all the triggers for a single resource, for example).  We'll need to find the sweet spot between function and efficiency/practicality.